src/overlays/actors/ovl_Obj_Timeblock/z_obj_timeblock.c

```diff
@@ -27,41 +57,314 @@ const ActorInit Obj_Timeblock_InitVars = {
+        if (blockSize < fabsf(distance.x) || blockSize < fabsf(distance.z)) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+s32 ObjTimeblock_WaitForOcarina(ObjTimeblock* this, GlobalContext* globalCtx) {
+    Player* player = PLAYER;
+
+    if (ObjTimeblock_PlayerIsInRange(this, globalCtx)) {
+        if (player->stateFlags2 & 0x1000000) {
+            func_8010BD58(globalCtx, 1);
+            this->songObserverFunc = &ObjTimeblock_WaitForSong;
+        } else {
+            player->stateFlags2 |= 0x800000;
+        }
+    }
+    return false;
+}
+
+s32 ObjTimeblock_WaitForSong(ObjTimeblock* this, GlobalContext* globalCtx) {
+    if (globalCtx->msgCtx.unk_E3EE == 4) {
+        this->songObserverFunc = &ObjTimeblock_WaitForOcarina;
+    }
+    if (globalCtx->msgCtx.unk_E3EC == 10) {
+        if (this->unk_172 == 254) {
+            this->songEndTimer = 110;
+        } else {
+            this->songEndTimer--;
+            if (this->songEndTimer == 0) {
+                return true;
+            }
+        }
+    }
+    return false;
+}
+
+void ObjTimeblock_StartDoNothingActionFunc(ObjTimeblock* this) {
+    this->actionFunc = &ObjTimeblock_DoNothingActionFunc;
+}
+
+void ObjTimeblock_DoNothingActionFunc(ObjTimeblock* this, GlobalContext* globalCtx) {
+}
+
+void ObjTimeblock_StartNormalActionFunc(ObjTimeblock* this) {
+    this->actionFunc = &ObjTimeblock_NormalActionFunc;
+}
+
+void ObjTimeblock_NormalActionFunc(ObjTimeblock* this, GlobalContext* globalCtx) {
+    u32 newIsVisible;
+
+    if (this->songObserverFunc(this, globalCtx) && this->demoEffectTimer <= 0) {
+        ObjTimeblock_SpawnDemoEffect(this, globalCtx);
+        this->demoEffectTimer = 160;
+
+        // Possibly points the camera to this actor
+        func_80080480(globalCtx, &this->dyna.actor);
+        // "◯◯◯◯ Time Block Attention Camera (frame counter  %d)\n"
+        osSyncPrintf("◯◯◯◯ Time Block 注目カメラ (frame counter  %d)\n", globalCtx->state.frames);
+
+        this->demoEffectFirstPartTimer = 12;
+
+        if (this->save == OBJ_TIMEBLOCK_DO_NOT_SAVE) {
+            this->dyna.actor.params ^= 0x8000;
+        } else {
+            ObjTimeblock_ToggleSwitchFlag(globalCtx, this->dyna.actor.params & 0x3F);
+        }
+    }
+
+    this->unk_172 = globalCtx->msgCtx.unk_E3EC;
+    if (this->demoEffectFirstPartTimer > 0) {
+        this->demoEffectFirstPartTimer--;
+        if (this->demoEffectFirstPartTimer == 0) {
+            if (this->save == OBJ_TIMEBLOCK_DO_NOT_SAVE) {
+                this->unk_175 = BOOL((this->dyna.actor.params >> 15) & 1);
+            } else {
+                this->unk_174 = BOOL(Flags_GetSwitch(globalCtx, this->dyna.actor.params & 0x3F));
+            }
+        }
+    }
+
+    newIsVisible = ObjTimeblock_CalculateIsVisible(this);
+    if (this->save == OBJ_TIMEBLOCK_SAVE && newIsVisible != this->isVisible) {
+        ObjTimeblock_StartDoNothingActionFunc(this);
```

Comment:
EDIT: I'm confused by how this actor handles its behavior, I'll still comment the following that I already wrote but idk

I think `save` is more of a `type`. For example here `OBJ_TIMEBLOCK_SAVE` would be "one-way only" or something
`OBJ_TIMEBLOCK_SAVE` also is used if the switch flag is >= 0x38 (if `(this->dyna.actor.params & 0x3F) < 0x38` is false), which is the "local" flag byte and is cleared "very often", so `SAVE` doesnt make a lot of sense
(see the [CloudModding wiki](https://wiki.cloudmodding.com/oot/Tutorial:Modification_and_Addition_of_Switches#Switch_Flags) (which usage should be avoided for decomp), [Flags_GetSwitch](https://github.com/zeldaret/oot/blob/82968a7381dc9e851487175174ea19750f060beb/src/code/z_actor.c#L523) and [func_80031B14](https://github.com/zeldaret/oot/blob/82968a7381dc9e851487175174ea19750f060beb/src/code/z_actor.c#L2503))